include/mqtt/broker/security.hpp

```diff
@@ -0,0 +1,492 @@
+    static inline std::string hash(std::string const &message)
+    {
+        std::shared_ptr<EVP_MD_CTX> mdctx(EVP_MD_CTX_new(), EVP_MD_CTX_free);
+        EVP_DigestInit_ex(mdctx.get(), EVP_sha256(), NULL);
+        EVP_DigestUpdate(mdctx.get(), message.data(), message.size());
+
+        std::vector<unsigned char> digest(static_cast<std::size_t>(EVP_MD_size(EVP_sha256())));
+        unsigned int digest_size = static_cast<unsigned int>(digest.size());
+
+        EVP_DigestFinal_ex(mdctx.get(), digest.data(), &digest_size);
+        return to_hex(digest.data(), digest.data() + digest_size);
+    }
+#else
+    static inline std::string hash(std::string const &message)
+    {
+        return message;
+    }
+#endif
+
+    bool login_cert(string_view const& username) const {
+        auto i = authentication_.find(std::string(username));
+        return (i != authentication_.end() && i->second.method_ == security::authentication::method::client_cert);
+    }
+
+    optional<std::string> login(string_view const& username, string_view const& password) const {
+        optional<std::string> empty_result;
+        auto i = authentication_.find(std::string(username));
+        if (i == authentication_.end() || i->second.method_ != security::authentication::method::password)
+            return empty_result;
+        return boost::iequals(*(i->second.password), hash(hash_type + ":" + salt + ":" + std::string(password))) ? std::string(username) : empty_result;
+    }
+
+    static authorization::type get_auth_type(string_view const& type) {
+        if(type == "allow") return authorization::type::allow;
+        if(type == "deny") return authorization::type::deny;
+        throw std::runtime_error("An invalid authorization type was specified: " + std::string(type));
+    }
+
+    static bool is_valid_group_name(string_view const& name) {
+        return !name.empty() && name[0] == '@'; // TODO: validate utf-8
+    }
+
+    static bool is_valid_user_name(string_view const& name) {
+        return !name.empty() && name[0] != '@'; // TODO: validate utf-8
+    }
+
+    void default_config()
+    {
+        hash_type = "aes256";
+        salt = "salt";
+
+        char const *username = "anonymous";
+        authentication login(authentication::method::anonymous);
+        authentication_.insert({ username, login});
+        anonymous = username;
+
+        char const *topic = "#";
+        authorization auth;
+        auth.topic_tokens = get_topic_filter_tokens("#");
+        auth.sub_type = authorization::type::allow;
+        auth.sub.insert(username);
+        auth.pub_type = authorization::type::allow;
+        auth.pub.insert(username);
+        authorization_.insert({ topic, auth });
+
+        validate();
+    }
+
+    void load_json(std::istream& input) {
+        // Create a root
+        boost::property_tree::ptree root;
+        boost::property_tree::read_json(input, root);
+
+        for(auto const &i: root.get_child("authentication")) {
```

Comment:
```suggestion
        for (auto const& i: root.get_child("authentication")) {
```

Please fix other same cases.